garrysmod/lua/includes/modules/outline.lua

```diff
@@ -0,0 +1,124 @@
+OUTLINE_MODE_BOTH,OUTLINE_MODE_NOTVISIBLE,OUTLINE_MODE_VISIBLE = 0,1,2
+
+module("outline",package.seeall)
+
+local List,ListSize = {},0
+local RenderEnt = NULL
+
+local CopyMat		= Material("pp/copy")
+local OutlineMat	= CreateMaterial("OutlineMat","UnlitGeneric",{["$ignorez"] = 1,["$alphatest"] = 1})
+local StoreTexture	= render.GetScreenEffectTexture(0)
+local DrawTexture	= render.GetScreenEffectTexture(1)
+
+local ENTS,COLOR,MODE = 1,2,3
+
+function Add(ents,color,mode)
+	if ListSize>=255 then return end		--Maximum 255 reference values
+	if !istable(ents) then ents = {ents} end	--Support for passing Entity as first argument
+	if ents[1]==nil then return end			--Do not pass empty tables
+	
+	table.insert(List,{
+		[ENTS] = ents,
+		[COLOR] = color,
+		[MODE] = mode or OUTLINE_MODE_BOTH
+	})
+	
+	ListSize = ListSize+1
+end
+
+function RenderedEntity()
+	return RenderEnt
+end
+
+local function Render()
+	local scene = render.GetRenderTarget()
+	render.CopyRenderTargetToTexture(StoreTexture)
+	
+	local w,h = ScrW(),ScrH()
+	
+	render.Clear(0,0,0,0,true,true)
+
+	render.SetStencilEnable(true)
+		cam.IgnoreZ(true)
+		render.SuppressEngineLighting(true)
+	
+		render.SetStencilWriteMask(255)
+		render.SetStencilTestMask(255)
+		
+		render.SetStencilCompareFunction(STENCIL_ALWAYS)
+		render.SetStencilFailOperation(STENCIL_KEEP)
+		render.SetStencilZFailOperation(STENCIL_REPLACE)
+		render.SetStencilPassOperation(STENCIL_REPLACE)
+		
+		cam.Start3D()
+			for k=1,ListSize do
+				local v = List[k]
+				local mode = v[MODE]
+				
+				render.SetStencilReferenceValue(k)
+				
+				for k2=1,#v[ENTS] do
+					local v2 = v[ENTS][k2]
+					
+					if !IsValid(v2) then continue end
+					
+					local visible = LocalPlayer():IsLineOfSightClear(v2)
```

Comment:
Since you now have OUTLINE_MODE_BOTH, you can skip this trace if the mode is not OUTLINE_MODE_NOTVISIBLE nor OUTLINE_MODE_VISIBLE.

Reply:
I think about it, but what if mode value is not equal to `both`,`notvisible`,`visible`, now there with checks used `both`, if not `notvisible` and not `visible`.

Reply:
I think you should treat all non-``nonvisible`` and ``visible`` values as both.